code/game/objects/items/glassjar.dm

```diff
@@ -31,75 +33,109 @@
+				usr.visible_message("<b>[usr]</b> takes a gumball from \the [src].", SPAN_NOTICE("You take a gumball from \the [src]."))
+				if(length(gumballs_contained) == 0)
+					contains = JAR_NOTHING
+				update_icon()
+
 
 /obj/item/glass_jar/afterattack(var/atom/A, var/mob/user, var/proximity)
 	if(!proximity || contains)
 		return
 	if(istype(A, /obj/effect/spider/spiderling))
 		var/obj/effect/spider/spiderling/S = A
-		user.visible_message(SPAN_NOTICE("<b>\The [user]</b> scoops \the [S] into \the [src]."), SPAN_NOTICE("You scoop \the [S] into \the [src]."))
-		S.forceMove(src)
 		STOP_PROCESSING(SSprocessing, S)	// No growing inside jars
 		contains = JAR_SPIDERLING
-		update_icon()
-		return
+		scoop(S, user)
 	if(ismob(A))
 		var/mob/L = A
 		if(L.mob_size <= MOB_TINY)
-			user.visible_message(SPAN_NOTICE("<b>\The [user]</b> scoops \the [L] into \the [src]."), SPAN_NOTICE("You scoop \the [L] into \the [src]."))
-			L.forceMove(src)
 			contains = JAR_ANIMAL
-			update_icon()
-			return
+			scoop(L, user)
 		else
-			to_chat(user, SPAN_WARNING("\The [L] doesn't fit into \the [src]!"))
-			return
+			scoop_fail(L, user)
+
+/obj/item/glass_jar/proc/scoop(var/atom/movable/A, var/mob/user)
+	user.visible_message(SPAN_NOTICE("<b>[user]</b> scoops \the [A] into \the [src]."), SPAN_NOTICE("You scoop \the [A] into \the [src]."))
+	playsound(src, pickup_sound, PICKUP_SOUND_VOLUME)
+	A.forceMove(src)
+	update_icon()
+	return
+
+/obj/item/glass_jar/proc/scoop_fail(var/atom/A, var/mob/user)
+	to_chat(user, SPAN_WARNING("\The [A] doesn't fit into \the [initial(name)]!"))
+	playsound(src, drop_sound, DROP_SOUND_VOLUME)
+	return
 
 /obj/item/glass_jar/attack_self(var/mob/user)
 	switch(contains)
 		if(JAR_MONEY)
 			for(var/obj/O in src)
 				O.forceMove(user.loc)
-			to_chat(user, SPAN_NOTICE("You take money out of \the [src]."))
-			contains = JAR_NOTHING
-			update_icon()
-			return
+				release(O, user)
 		if(JAR_ANIMAL)
 			for(var/mob/M in src)
 				M.forceMove(user.loc)
-				user.visible_message(SPAN_NOTICE("\The [user] releases \the [M] from \the [src]."), SPAN_NOTICE("You release \the [M] from \the [src]."))
-			contains = JAR_NOTHING
-			update_icon()
-			return
+				release(M, user)
 		if(JAR_SPIDERLING)
 			for(var/obj/effect/spider/spiderling/S in src)
 				S.forceMove(user.loc)
-				user.visible_message(SPAN_NOTICE("\The [user] releases \the [S] from \the [src]."), SPAN_NOTICE("You release \the [S] from \the [src]."))
 				START_PROCESSING(SSprocessing, S) // They can grow after being let out though
-			contains = JAR_NOTHING
-			update_icon()
-			return
+				release(S, user)
+		if(JAR_HOLDER)
+			for(var/obj/item/holder/H in src)
+				H.release_to_floor() // Snowflake code because holders are ass. Q.E.D.
+				release(H, user)
 		if(JAR_NOTHING)
 			to_chat(user, SPAN_NOTICE("You remove the lid from \the [src]."))
 			user.drop_from_inventory(src)
 			user.put_in_hands(new /obj/item/reagent_containers/glass/beaker/jar) //found in jar.dm
```

Comment:
you need to make it so the jar is dropped from your inventory in `jar.dm` between lines 21 and 22
otherwise it swaps hands every time you put the lid back on

(i'm also really not a fan of how it suddenly turns into a reagent container if you take the lid off,
and how you can only take the lid off if there's nothing inside)

Reply:
Well, there's a lot of interactions with reagent containers which is a whole can of worms I honestly don't want to deal with.

Jar.dm has been fixed